jekyll/_cci2/caching.md

```diff
@@ -1,74 +1,172 @@
+
+The Docker image(s) you use for your jobs will usually have tools preinstalled that are generic for building projects using the language the image is focused on. For example the `circleci/ruby:2.4.1` image has useful tools like git, openssh-client and gzip preinstalled. So these types of tools that are not explicitly required for your project are good candidates for being stored on the Docker image.
+
+## How do I know what the path is for my dependencies?
+
+The various language dependency managers (e.g. `npm` or `pip`) each have their own paths where dependencies are installed. See our Language guides and demo projects for the specifics for your stack: <https://circleci.com/docs/2.0/demo-apps/>
+
+## Using cache keys and template values to control what is cached and restored
+
+This section is where you will learn about the art of caching. We say this because caching is a balance between getting maximum performance (using a full cache for every build) and reliability (not using an out-of-date or inappropriate cache).
+
+The way you decide on when a cache will be save or restored is to use a key which is tied an aspect of your project.
+
+When storing a new cache, the `key` value may contain special template values:
+
+Template | Description
+----|----------
+{% raw %}`{{ .Branch }}`{% endraw %} | the VCS branch currently being built
+{% raw %}`{{ .BuildNum }}`{% endraw %} | the CircleCI build number for this build
+{% raw %}`{{ .Revision }}`{% endraw %} | the VCS revision currently being built
+{% raw %}`{{ .Environment.variableName }}`{% endraw %} | the environment variable `variableName`
+{% raw %}`{{ checksum "filename" }}`{% endraw %} | a base64 encoded SHA256 hash of the given filename's contents. This should be a file committed in your repo. Good candidates are dependency manifests, such as `package.json`, `pom.xml` or `project.clj`. It's important that this file does not change between `restore_cache` and `save_cache`, otherwise the cache will be saved under a cache key different than the one used at `restore_cache` time.
+{% raw %}`{{ epoch }}`{% endraw %} | the current time in seconds since the unix epoch.
+{: class="table table-striped"}
+
+During step execution, the templates above will be replaced by runtime values and use the resultant string as the `key`.
+
+### Choose a strategy by understanding how cache is restored
+
+To decide how to save your cache, it's useful to first understand how CircleCI selects what will be restored. Here's an example of a restore step using multiple keys:
 
-try:
 {% raw %}
 ```YAML
 - restore_cache:
     keys:
-      - projectname-{{ .Branch }}-{{ checksum "package.json" }}
-      - projectname-{{ .Branch }}
-      - projectname-master
+      # Find a cache corresponding to this specific package.json checksum
+      # when this file is changed, this key will fail
+      - v1-projectname-npm-deps-{{ .Branch }}-{{ checksum "package.json" }}
```

Comment:
ok, I removed the second item in the list,  removed branch from the first item, and removed project name, let me know if that is not correct